### ISSUE TITLE
Scripts not running when switching home-manager configuration

### ISSUE THREAD

**Issue by HeitorAugustoLN** [nix-community/plasma-manager#162]
I've been trying to add the panel spacer to my panel with plasma-manager, but it is not working. I tried rebooting, but nothing so far.

This is what my plasma-manager config looks like
![image](https://github.com/pjones/plasma-manager/assets/44377258/b56256ca-4155-4c4e-bc50-e5d16c0fbb42)

This is how my panel is looking like:

![image](https://github.com/pjones/plasma-manager/assets/44377258/4ca7304f-fa16-48f5-a0b8-8e38194ea05f)

**Comment by magnouvean** [comment]
I assume you want the icontasks in the center? I tried the following:
```nix
widgets = [
  "org.kde.plasma.panelspacer"
  "org.kde.plasma.icontasks";
  "org.kde.plasma.panelspacer"
];
```
and it works fine so I can't really reproduce (I'm on plasma 6 though). If you add them manually at the locations you want does it work then? If it is a case of them not being added, what is the content in `~/.local/share/plasma-manager/data/desktop_script_apply_panels.js`? From what I can tell `plasma-manager` handles this as expected. Also make sure you are on the latest version :)

**Comment by HeitorAugustoLN** [comment]
I tried changing to what you've done, but not working. And i maybe found out why, my panel settings aren't updating, in the interface, only in the script

This is the content before I tried your setting:
![image](https://github.com/pjones/plasma-manager/assets/44377258/7b40d471-0501-4327-b939-1ed3f62bd62d)

After (my current setting):
![image](https://github.com/pjones/plasma-manager/assets/44377258/292000ab-854f-4746-a2ec-0bcdc22cf442)

This is how my panel is looking like:
![image](https://github.com/pjones/plasma-manager/assets/44377258/a3cb6b0b-d960-4148-95a9-6cb89152c00c)

**Comment by HeitorAugustoLN** [comment]
I will try deleting the panel, then switching my config, and rebooting to see if it is solved

**Comment by HeitorAugustoLN** [comment]
> I will try deleting the panel, then switching my config, and rebooting to see if it is solved

didnt work lol

**Comment by magnouvean** [comment]
The panel should be deleted and regenerated on each login (at least when the configuration has changed) so that's probably not the issue. The `js` file looks fine, I don't really know what's happening. What plasma version are you running?

**Comment by magnouvean** [comment]
Actually looking at the panel generated that looks to be the default panel as there are widgets there not specified by the config at all. What happens if you run
```
rm ~/.local/share/plasma-manager/last_run_*
~/.local/share/plasma-manager/run_all.sh
```
Is there any output in console after this?

**Comment by HeitorAugustoLN** [comment]
Now my panel is updated, this was the output:

![image](https://github.com/pjones/plasma-manager/assets/44377258/5d88af61-ff7e-4390-9bbd-6d27f290c4d0)

**Comment by HeitorAugustoLN** [comment]
> Now my panel is updated, this was the output:
> 
> ![image](https://private-user-images.githubusercontent.com/44377258/327824520-5d88af61-ff7e-4390-9bbd-6d27f290c4d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3NTYzNDYsIm5iZiI6MTcxNDc1NjA0NiwicGF0aCI6Ii80NDM3NzI1OC8zMjc4MjQ1MjAtNWQ4OGFmNjEtZmY3ZS00MzkwLTliYmQtNmQyN2YyOTBjNGQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDE3MDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2OGY5ZmJkZjM3NDhhMWFlM2EwNTE2ZGNiM2Q4NDhhM2MxYjMxZDM4YTdlMWM0YzRmMjNmNmI2MjIzZWU4M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-00kb7lcDSOMmhFq5RQlvRpJR7V4Q36m_0ADhhbZVAI)

Translated output:

The theme of the current Plasma session has been set to breeze-dark
The “Breeze” theme could not be found. The theme can only be one of the following: breeze_cursors, Breeze_Light
The requested theme “BreezeDark” is already set as the theme for the current Plasma session.

The definition of a wallpaper for all desktops in the image /nix/store/sz49ipfjcj2vk94wya31861lzr999a9i-source/93.jpg has been successfully completed.

Translated with DeepL.com (free version)

**Comment by HeitorAugustoLN** [comment]
> Now my panel is updated, this was the output:
> 
> ![image](https://private-user-images.githubusercontent.com/44377258/327824520-5d88af61-ff7e-4390-9bbd-6d27f290c4d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ3NTYzNDYsIm5iZiI6MTcxNDc1NjA0NiwicGF0aCI6Ii80NDM3NzI1OC8zMjc4MjQ1MjAtNWQ4OGFmNjEtZmY3ZS00MzkwLTliYmQtNmQyN2YyOTBjNGQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAzVDE3MDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2OGY5ZmJkZjM3NDhhMWFlM2EwNTE2ZGNiM2Q4NDhhM2MxYjMxZDM4YTdlMWM0YzRmMjNmNmI2MjIzZWU4M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-00kb7lcDSOMmhFq5RQlvRpJR7V4Q36m_0ADhhbZVAI)

Translated output:

The theme of the current Plasma session has been set to breeze-dark
The “Breeze” theme could not be found. The theme can only be one of the following: breeze_cursors, Breeze_Light
The requested theme “BreezeDark” is already set as the theme for the current Plasma session.

The definition of a wallpaper for all desktops in the image /nix/store/sz49ipfjcj2vk94wya31861lzr999a9i-source/93.jpg has been successfully completed.

**Comment by magnouvean** [comment]
From the outputs there is no indications of problems on the panels side. The problem doesn't look to be with `org.kde.plasma.panelspacer` to me, but there may be problems with the scripts not running when they should (removing the last_run_* files ensures the panel generation script is run at next login). When using `overrideConfig` this hasn't been an issue as far as I can tell, hence why I wasn't able to reproduce. I'll see if I can investigate a bit further some time and see if I can reproduce. If I'm not able to I'm out of ideas.

**Comment by HeitorAugustoLN** [comment]
> From the outputs there is no indications of problems on the panels side. The problem doesn't look to be with `org.kde.plasma.panelspacer` to me, but there may be problems with the scripts not running when they should (removing the last_run_* files ensures the panel generation script is run at next login). When using `overrideConfig` this hasn't been an issue as far as I can tell, hence why I wasn't able to reproduce. I'll see if I can investigate a bit further some time and see if I can reproduce. If I'm not able to I'm out of ideas.

Yeah, i noted this, i will change the issue title.

**Comment by NovaViper** [comment]
I'm having the same issue as well too. The script for the panels specifically doesn't run while everything else does when I use switch home-manager. However when I run the `run-all.sh` script, it applies all of the settings directly
```
  programs = {
    konsole = {
      enable = mkIf (config.variables.useKonsole) true;
      defaultProfile = "DefaultThemed";
      /*
        profiles.DefaultThemed = {
        name = "DefaultThemed";
        colorScheme = "Dracula";
        font = {
          name = "${config.stylix.fonts.monospace.name}";
          size = config.stylix.fonts.sizes.terminal;
        };
      };
      */
    };
    plasma = {
      enable = true;
      workspace.clickItemTo = mkDefault "select";
      kwin.titlebarButtons = {
        left = ["on-all-desktops" "keep-above-windows"];
        right = ["help" "minimize" "maximize" "close"];
      };
      panels = [
        # Windows like panel at the bottom
        {
          location = "bottom";
          height = 46;
          floating = false;
          widgets = [
            {
              name = "org.kde.plasma.kickoff";
              config = {
                General.icon = "nix-snowflake-white";
              };
            }
            "org.kde.plasma.marginsseparator"
            {
              name = "org.kde.plasma.icontasks";
              /*
                config = {
                Genera.launchers = [
                  ""
                ];
              };
              */
            }
            "org.kde.plasma.marginsseparator"
            "org.kde.plasma.pager"
            "org.kde.plasma.showdesktop"
          ];
        }
        {
          location = "top";
          height = 26;
          widgets = [
            "org.kde.plasma.appmenu"
            {
              digitalClock = {
                calendar.firstDayOfWeek = "sunday";
                date = {
                  enable = true;
                  position = "besideTime";
                };
                time.showSeconds = "always";
              };
            }
            {
              systemTray = {
                icons.scaleToFit = true;
                items.configs.battery.showPercentage = true;
              };
            }
          ];
        }
      ];
      shortcuts.yakuake =
        mkIf config.variables.useKonsole {toggle-window-state = "F12";};
      configFile = {
        kglobalshortcutsrc.yakuake = mkIf config.variables.useKonsole {
          "_k_friendly_name".value = "Yakuake";
        };
        kwinrc.NightColor = {
          Active.value = true;
          EveningBeginFixed = 2000;
          LatitudeAuto = "30.51";
          LongitudeAuto = "-91.12";
          Mode = "Times";
        };
        kservicemenurc.Show = {
          "compressfileitemaction" = true;
          "extractfileitemaction" = true;
          "forgetfileitemaction" = true;
          "installFont" = true;
          "kactivitymanagerd_fileitem_linking_plugin" = true;
          "kdeconnectfileitemaction" = true;
          "kio-admin" = true;
          "makefileactions" = true;
          "mountisoaction" = true;
          "plasmavaultfileitemaction" = true;
          "runInKonsole" = true;
          "slideshowfileitemaction" = true;
          "tagsfileitemaction" = true;
          "wallpaperfileitemaction" = true;
        };
        dolphinrc = {
          ContentDisplay.UsePermissionsFormat = "CombinedFormat";
          VersionControl.enabledPlugins = "Git";
          General = {
            "AutoExpandFolders" = true;
            "FilterBar" = true;
            "ShowFullPathInTitlebar" = true;
            "ShowToolTips" = true;
            "SortingChoice" = "CaseInsensitiveSorting";
            "UseTabForSwitchingSplitView" = true;
          };
          "KFileDialog Settings" = {
            "Places Icons Auto-resize" = false;
            "Places Icons Static Size" = 22;
          };
          PreviewSettings.Plugins = "appimagethumbnail,audiothumbnail,blenderthumbnail,comicbookthumbnail,cursorthumbnail,djvuthumbnail,ebookthumbnail,exrthumbnail,directorythumbnail,fontthumbnail,imagethumbnail,jpegthumbnail,kraorathumbnail,windowsexethumbnail,windowsimagethumbnail,mobithumbnail,opendocumentthumbnail,gsthumbnail,rawthumbnail,svgthumbnail,ffmpegthumbs";
        };
      };
    };
  };
```

home-manager switch log
```
> Building home configuration
Using saved setting for 'extra-substituters = https://nix-community.cachix.org https://nix-gaming.cachix.org' from ~/.local/share/nix/trusted-settings.json.
Using saved setting for 'extra-trusted-public-keys = nix-community.cachix.org-1:mB9FSh9qf2dCimDSUo8Zy7bkq5CX+/rkCWyvRCYg3Fs= nix-gaming.cachix.org-1:nbjlureqMbRAxR1gJ/f3hxemL9svXaZF/Ees8vCUUs4=' from ~/.local/share/nix/trusted-settings.json.
Finished at 12:55:03 after 16s
> Comparing changes
<<< /home/novaviper/.local/state/nix/profiles/home-manager
>>> /tmp/nh-home-KW6K1s/result
No version or selection state changes.
Closure size: 3676 -> 3676 (0 paths added, 0 paths removed, delta +0, disk usage +0B).
> Activating configuration
Starting Home Manager activation
Activating checkFilesChanged
Activating checkLinkTargets
Activating writeBoundary
Activating createGpgHomedir
Activating createMaildir
Activating linkGeneration
Cleaning up orphan links from /home/novaviper
No change so reusing latest profile generation 78
Creating home file links in /home/novaviper
Activating batCache
No syntaxes were found in '/home/novaviper/.config/bat/syntaxes', using the default set.
Writing theme set to /home/novaviper/.cache/bat/themes.bin ... okay
Writing syntax set to /home/novaviper/.cache/bat/syntaxes.bin ... okay
Writing metadata to folder /home/novaviper/.cache/bat ... okay
Activating configure-plasma
Activating createXdgUserDirectories
Activating installPackages
replacing old 'home-manager-path'
installing 'home-manager-path'
Activating dconfSettings
Activating installDoomEmacs
Activating onFilesChange
Activating reloadSystemd
Activating runMuInit
Activating stylixLookAndFeel
Successfully set the wallpaper for all desktops to the KPackage based /nix/store/gsi2pk7wbc0q2nmy8vwmwj0gw59j76dr-stylix-kde-theme/share/wallpapers/stylix
KPackageStructure of KPluginMetaData(pluginId:"stylix", fileName: "/home/novaviper/.nix-profile/share/plasma/look-and-feel/stylix/metadata.json") does not match requested format "Plasma/LookAndFeel"
```

**Comment by magnouvean** [comment]
That sounds very weird. Are you saying that running `~/.local/share/plasma-manager/run_all.sh` applies everything fine, but that the scripts doesn't work on login? That would be very weird since plasma-manager just runs that script automatically on login through an autostart entry. The only possible explanation I could make of that would be if the script is run too early or something, but I have never experienced that to be an issue. Is this plasma 6?

**Comment by NovaViper** [comment]
@magnouvean Yeah I can run the script manually but if I use home-manager to switch the profiles and then restart, it doesn't run at all.

Also you mention an autostart entry, I see it in the autostart page in the KDE settings tab.. but when I hit to run it just doesn't do anything, just says running for a second and then just instantly stops.

**Comment by HeitorAugustoLN** [comment]
> That sounds very weird. Are you saying that running `~/.local/share/plasma-manager/run_all.sh` applies everything fine, but that the scripts doesn't work on login? That would be very weird since plasma-manager just runs that script automatically on login through an autostart entry. The only possible explanation I could make of that would be if the script is run too early or something, but I have never experienced that to be an issue. Is this plasma 6?

Yeah, this is what basically is happening, when I change my panel settings, I need to run the script manually, and then it works. And since when I opened this issue, i've changed from a standalone home-manager configuration to a nixos module configuration. And still having the same problem

**Comment by HeitorAugustoLN** [comment]
But this only happens after you first configure the panel via plasma-manager, and then modify it

**Comment by magnouvean** [comment]
I see that the `sha256sum` comparison for plasma-scripts in the `startup` module may have some problems as the script which runs at startup doesn't change between activations, only the `layout.js` files does that (possibly causing the panel to only be executed once, without re-executing the script when changes are made to the panels). I can create a fork with some tweaks to this and we can see if that solves the problem.

If that isn't the problem it's probably some kind of problem with how early the script is executed, which is a bit harder to solve. In that case maybe we need to explore using systemd-services instead of autostart scripts (maybe this should be done eventually anyway). If even this fails we could just run the script on every login, even if no changes have been made (hardly optimal, but better than being broken as it is now at the very least).

**Comment by magnouvean** [comment]
I have tested this in the [scripts-sha256sum-fix in my fork](https://github.com/magnouvean/plasma-manager/tree/scripts-sha256sum-fix). Does the problem still persist if you use this as the plasma-manager source (that is if you set `github:magnouvean/plasma-manager/scripts-sha256sum-fix` as the plasma-manager url in the `flakes.nix`).

**Comment by NovaViper** [comment]
@magnouvean So far so good. I can use that Autostart KCM page to actually start the service and it actually runs instead of instantly stopping like it was before 👍🏾 Will update when my computer comes back on 

Edit: Huh.. it broke again after I restarted my computer. 🤔 I do have Home-manager in both standalone and NixOS module mode so going to do another quick to see maybe it'll make a difference

Edit 2: It looks like it works now after changing the panel layout and rebooting the computer! The script is suppose to only execute once when it detects changes correct?

**Comment by magnouvean** [comment]
Yes @NovaViper, the script is meant to run only when there is a change in the panel configuration, and then not run again until there is a change in some panel in some way. Does it regenerate the panels if you change the panel config in any way (change the height of a panel for example)? Additionally if you don't make any changes to the panel config, does the script then not run? If both of those are yes then this seems to be working :)

**Comment by NovaViper** [comment]
@magnouvean Hey sorry for the delay! 🤔 Looks like it's working properly but there's this sort of weird quirk going on with home-manager and NixOS.  It does try to switch when I boot up the computer but it goes back to the old layout before I ran the switch command for Home-Manager. However when I run `nixos-rebuild switch`, and then reboot, it runs the script and then loads up the new configs. That I think is a quirk with how I have Home-Manger both as standalone and being ran as a NixOS module.

**Comment by magnouvean** [comment]
Yeah that's probably not related to this (usually one uses either the module, standalone or the flake home-manager) 😀. Seeing as this seems to work fine (and fixes an important bug in the project) I will make a pr and merge soon, and then I'll leave it up to @HeitorAugustoLN to close the issue if it works there as well 😄

**Comment by HeitorAugustoLN** [comment]
After #179 scripts are running as normal. But when I was testing I realized another bug, sometimes when switching settings, my wallpaper gets reset to Plasma default wallpaper. I tested it 4 times, and it got reset 2 times. Couldn't figure what was causing it. But the panels and everything else is updating just fine

**Comment by magnouvean** [comment]
It's probably due to the panel setup script removing `plasma-org.kde.plasma.desktop-appletsrc` which also is where the wallpaper is set, so running the panel script without running the wallpaper script could result in the default wallpaper. The reason for removing the file was that the file started to grow quite a lot (possibly due to some plasma bug), but maybe we should reconsider that. This shouldn't be related to the current issue though.

**Comment by HeitorAugustoLN** [comment]
Ok, so I'll proceed with closing this issue. To address the new problem, should I open a new issue?

**Comment by magnouvean** [comment]
I plan on making some smaller changes today, so we can see if it's fixed by that. I can notify here, and if it's not solved by that you can open a new issue :)

**Comment by magnouvean** [comment]
I have made some changes in [the wallpaper-panels-script-tweak branch in my fork](https://github.com/magnouvean/plasma-manager/tree/wallpaper-panels-script-tweak). See if setting the plasma-manager url to `github:magnouvean/plasma-manager/wallpaper-panels-script-tweak` fixes the above problem. If not I reckon open a separate issue @HeitorAugustoLN :)

**Comment by HeitorAugustoLN** [comment]
As soon as I get home i will test it

-------- Mensagem original --------
Em 01/06/2024 10:09, magnouvean  escreveu:

**Comment by HeitorAugustoLN** [comment]
@magnouvean tested it here, it is working fine. No wallpaper reset whatsoever.

**Comment by magnouvean** [comment]
Nice. I want to test a few things before merging, so it might take a few days before it reaches the trunk branch though :)
